File: packages/guarddog-ios-sdk/GuardDogCore/Sources/GuardDogCore/Events/SecurityEvent.swift
import Foundation

/// Mirrors packages/guarddog-contracts/src/securityEvent.ts. Expo-independent.
public enum SecurityEventType: String, Codable {
    case threatBlocked = "THREAT_BLOCKED"
    case threatDetected = "THREAT_DETECTED"
    case protectionStateChanged = "PROTECTION_STATE_CHANGED"
    case ruleBundleAccepted = "RULE_BUNDLE_ACCEPTED"
    case ruleBundleRejected = "RULE_BUNDLE_REJECTED"
}

public enum SecurityEventSource: String, Codable {
    case androidVpnEnforcement = "android-vpn-enforcement"
    case localAnalysis = "local-analysis"
    case ruleVerifier = "rule-verifier"
    case protectionLifecycle = "protection-lifecycle"
}

public struct SecurityEvent: Codable, Equatable {
    public var id: String
    public var type: SecurityEventType
    public var source: SecurityEventSource
    public var occurredAt: String
    public var sanitizedUrl: String?
    public var host: String?
    public var destinationIp: String?
    public var ruleId: String?
    public var rulesetId: String?
    public var bundleVersion: Int64?
    public var enforcementEvidenceId: String?
    public var verdict: String?
    public var protectionState: String?
    public var reason: String?

    public init(id: String, type: SecurityEventType, source: SecurityEventSource, occurredAt: String, sanitizedUrl: String? = nil, host: String? = nil,
                destinationIp: String? = nil, ruleId: String? = nil, rulesetId: String? = nil, bundleVersion: Int64? = nil,
                enforcementEvidenceId: String? = nil, verdict: String? = nil, protectionState: String? = nil, reason: String? = nil) {
        self.id = id; self.type = type; self.source = source; self.occurredAt = occurredAt; self.sanitizedUrl = sanitizedUrl; self.host = host
        self.destinationIp = destinationIp; self.ruleId = ruleId; self.rulesetId = rulesetId; self.bundleVersion = bundleVersion
        self.enforcementEvidenceId = enforcementEvidenceId; self.verdict = verdict; self.protectionState = protectionState; self.reason = reason
    }

    /// iOS never produces this in M1: there is no enforcement layer. Kept for contract parity.
    public var isGenuineBlock: Bool {
        type == .threatBlocked && source == .androidVpnEnforcement && !(enforcementEvidenceId ?? "").isEmpty
    }
}

File: packages/guarddog-android-sdk/guarddog-core/src/main/java/com/guarddog/core/events/SecurityEvent.kt
package com.guarddog.core.events

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Mirrors packages/guarddog-contracts/src/securityEvent.ts. Expo-independent. */
@Serializable
enum class SecurityEventType {
    @SerialName("THREAT_BLOCKED") THREAT_BLOCKED,
    @SerialName("THREAT_DETECTED") THREAT_DETECTED,
    @SerialName("PROTECTION_STATE_CHANGED") PROTECTION_STATE_CHANGED,
    @SerialName("RULE_BUNDLE_ACCEPTED") RULE_BUNDLE_ACCEPTED,
    @SerialName("RULE_BUNDLE_REJECTED") RULE_BUNDLE_REJECTED,
}

@Serializable
enum class SecurityEventSource {
    @SerialName("android-vpn-enforcement") ANDROID_VPN_ENFORCEMENT,
    @SerialName("local-analysis") LOCAL_ANALYSIS,
    @SerialName("rule-verifier") RULE_VERIFIER,
    @SerialName("protection-lifecycle") PROTECTION_LIFECYCLE,
}

@Serializable
data class SecurityEvent(
    val id: String,
    val type: SecurityEventType,
    val source: SecurityEventSource,
    /** ISO-8601 UTC, e.g. 2026-06-15T00:00:00Z */
    val occurredAt: String,
    val sanitizedUrl: String? = null,
    val host: String? = null,
    val destinationIp: String? = null,
    val ruleId: String? = null,
    val rulesetId: String? = null,
    val bundleVersion: Long? = null,
    val enforcementEvidenceId: String? = null,
    val verdict: String? = null,
    val protectionState: String? = null,
    val reason: String? = null,
    // --- Additive cross-platform evidence fields (Gate Guard M2), mirrors securityEvent.ts 1:1. M1 events
    // never populate these; null is the correct, expected value for every M1 event, past or future. ---
    val platform: String? = null,
    val osVersion: String? = null,
    /** Apollo core/SDK engine version that produced this event. */
    val engineVersion: String? = null,
    /** e.g. "android-vpn-tun-drop" (M1) or "android-vpn-dns-sinkhole-drop" (M2 website gate). */
    val enforcementMechanism: String? = null,
    val direction: String? = null,
    /** What the decision layer asked for vs what enforcement actually did (equal for every M1 event). */
    val actionRequested: String? = null,
    val actionEnforced: String? = null,
    /** 0..1 provider/rule confidence, when the source supplies one. */
    val confidence: Double? = null,
    /** Explicit null (not merely absent) when the platform cannot provide this (Android/iOS without
     * AccessibilityService-class permissions, which Apollo does not use). */
    val applicationIdentity: String? = null,
) {
    /** True only for events produced from real packet-drop evidence. */
    val isGenuineBlock: Boolean
        get() = type == SecurityEventType.THREAT_BLOCKED &&
            source == SecurityEventSource.ANDROID_VPN_ENFORCEMENT &&
            !enforcementEvidenceId.isNullOrEmpty()
}
